# MLOps Pipeline for Hospital Readmission Classification using the UCI Diabetes Dataset

This project implements an end-to-end MLOps pipeline for predicting hospital readmission risk using the UCI Diabetes 130-US Hospitals dataset. The objective is to build a reliable machine learning classification system that identifies whether a patient is likely to be readmitted within 30 days based on historical clinical records.

The dataset contains over 100,000 patient records with 47 features, including demographic information, diagnoses, medication history, and hospital visit details. It presents real-world challenges such as missing values in multiple formats, high-cardinality categorical variables, and significant class imbalance.

The pipeline covers data ingestion, preprocessing, feature engineering, model training, and evaluation using appropriate metrics such as ROC-AUC, precision, and recall. The project is structured with an MLOps mindset to support reproducibility, scalability, and deployment readiness in a healthcare environment.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [24]:
data = pd.read_csv("diabetic_data.csv", na_values = "?")
data.sample(10)

C:\Users\HomePC\AppData\Local\Temp\ipykernel_16200\1229332509.py:1: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv("diabetic_data.csv", na_values = "?")


,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
73854,219164262,75419874,Caucasian,Male,[60-70),NaN,1,1,7,13,...,No,Down,No,No,No,No,No,Ch,Yes,NO
72265,211078620,105336846,AfricanAmerican,Female,[90-100),NaN,3,3,1,3,...,No,Down,No,No,No,No,No,Ch,Yes,>30
37777,117250170,87307578,Caucasian,Male,[50-60),NaN,3,1,1,8,...,No,Steady,No,No,No,No,No,Ch,Yes,NO
65039,181111422,27667008,Caucasian,Female,[80-90),NaN,1,1,7,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO
57190,163452270,73346715,Caucasian,Female,[70-80),NaN,1,3,7,8,...,No,Steady,No,No,No,No,No,No,Yes,NO
3169,19975548,23600844,Caucasian,Male,[90-100),NaN,5,11,17,1,...,No,Steady,No,No,No,No,No,No,Yes,NO
101134,435494330,86072490,AfricanAmerican,Female,[50-60),NaN,3,1,1,2,...,No,Down,No,No,No,No,No,Ch,Yes,NO
99205,410778380,103946688,Hispanic,Male,[70-80),NaN,3,1,1,3,...,No,No,No,No,No,No,No,Ch,Yes,NO
78124,238189326,74105955,Caucasian,Female,[60-70),NaN,3,1,1,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO
69971,199916598,84931803,Caucasian,Female,[70-80),NaN,1,1,7,3,...,No,No,No,No,No,No,No,Ch,Yes,NO


In [31]:
data.shape

(101766, 50)

In [30]:
missing = data.isnull().sum().sort_values(ascending=False)
missing_percent = (missing / len(data)) * 100

missing_df = pd.DataFrame({
    'missing_count': missing,
    'missing_percent': missing_percent
})

missing_df.head(10)

,missing_count,missing_percent
weight,98569,96.858479
max_glu_serum,96420,94.746772
A1Cresult,84748,83.277322
medical_specialty,49949,49.082208
payer_code,40256,39.557416
race,2273,2.233555
diag_3,1423,1.398306
diag_2,358,0.351787
diag_1,21,0.020636
encounter_id,0,0.000000


The dataset contains 101,766 patient encounters and 50 features. It is a real-world clinical dataset characterized by significant missing data, particularly in laboratory and specialty-related features. The level of missingness varies across features, requiring a combination of feature removal and targeted imputation strategies to ensure data quality for modeling.